File: ios/Packages/FatwaBotKit/Sources/CoreKit/ConfigModels.swift
import Foundation

// Server contract models — mirror backend/openapi/api.v1.yaml.

public struct AppConfig: Codable, Equatable, Sendable {
    public struct Flag: Codable, Equatable, Sendable {
        public struct Rollout: Codable, Equatable, Sendable {
            public let minAppVersion: String?

            enum CodingKeys: String, CodingKey {
                case minAppVersion = "min_app_version"
            }

            public init(minAppVersion: String? = nil) { self.minAppVersion = minAppVersion }
        }

        public let enabled: Bool
        public let rollout: Rollout?

        public init(enabled: Bool, rollout: Rollout? = nil) {
            self.enabled = enabled
            self.rollout = rollout
        }
    }

    public let config: [String: JSONValue]
    public let flags: [String: Flag]
    public let locales: [LocaleInfo]

    public init(config: [String: JSONValue], flags: [String: Flag], locales: [LocaleInfo]) {
        self.config = config
        self.flags = flags
        self.locales = locales
    }

    public func isEnabled(_ flag: String) -> Bool {
        flags[flag]?.enabled ?? false
    }
}

public struct LocaleInfo: Codable, Equatable, Sendable {

    public init(locale: String, displayName: String, direction: String, digits: String) {
        self.locale = locale
        self.displayName = displayName
        self.direction = direction
        self.digits = digits
    }
    public let locale: String
    public let displayName: String
    public let direction: String
    public let digits: String

    enum CodingKeys: String, CodingKey {
        case locale
        case displayName = "display_name"
        case direction
        case digits
    }
}

public struct ServerTheme: Codable, Equatable, Sendable {

    public init(version: Int, tokens: [String: JSONValue]) {
        self.version = version
        self.tokens = tokens
    }
    public let version: Int
    public let tokens: [String: JSONValue]
}

public struct StringPack: Codable, Equatable, Sendable {

    public init(locale: String, version: Int, strings: [String: String]) {
        self.locale = locale
        self.version = version
        self.strings = strings
    }
    public let locale: String
    public let version: Int
    public let strings: [String: String]
}

public struct HomeLayout: Codable, Equatable, Sendable {
    public struct Section: Codable, Equatable, Sendable {
        public init(id: String, type: String, props: [String: JSONValue]) {
            self.id = id
            self.type = type
            self.props = props
        }
        public let id: String
        public let type: String
        public let props: [String: JSONValue]
    }

    public let version: Int
    public let sections: [Section]

    public init(version: Int, sections: [Section]) {
        self.version = version
        self.sections = sections
    }

    /// ADR-0011 forward compatibility: unknown section types are skipped, never fatal.
    public func renderableSections(supported: Set<String>) -> [Section] {
        sections.filter { supported.contains($0.type) }
    }
}

public struct PrayerDefaults: Codable, Equatable, Sendable {

    public init(countryCode: String, method: String, params: [String: JSONValue]) {
        self.countryCode = countryCode
        self.method = method
        self.params = params
    }
    public let countryCode: String
    public let method: String
    public let params: [String: JSONValue]

    enum CodingKeys: String, CodingKey {
        case countryCode = "country_code"
        case method
        case params
    }
}

File: android/core/common/src/main/kotlin/com/fatwabot/core/common/ConfigModels.kt
package com.fatwabot.core.common

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonObject

// Server contract models — mirror backend/openapi/api.v1.yaml (parity with iOS CoreKit).

@Serializable
data class AppConfig(
    val config: JsonObject,
    val flags: Map<String, Flag>,
    val locales: List<LocaleInfo>,
) {
    @Serializable
    data class Flag(val enabled: Boolean, val rollout: Rollout? = null) {
        @Serializable
        data class Rollout(
            @SerialName("min_app_version") val minAppVersion: String? = null,
        )
    }

    fun isEnabled(flag: String): Boolean = flags[flag]?.enabled ?: false
}

@Serializable
data class LocaleInfo(
    val locale: String,
    @SerialName("display_name") val displayName: String,
    val direction: String,
    val digits: String,
)

@Serializable
data class ServerTheme(
    val version: Int,
    val tokens: JsonObject,
)

@Serializable
data class StringPack(
    val locale: String,
    val version: Int,
    val strings: Map<String, String>,
)

@Serializable
data class HomeLayout(
    val version: Int,
    val sections: List<Section>,
) {
    @Serializable
    data class Section(
        val id: String,
        val type: String,
        val props: JsonObject,
    )

    /** ADR-0011 forward compatibility: unknown section types are skipped, never fatal. */
    fun renderableSections(supported: Set<String>): List<Section> =
        sections.filter { it.type in supported }
}

@Serializable
data class PrayerDefaults(
    @SerialName("country_code") val countryCode: String,
    val method: String,
    val params: JsonElement,
)
